# Introduction 

In ecommerce companies like online retails, customer segmentation is necessary in order to understand customers behaviors. It leverages aqcuired customer data like the one we have in our case, **transactions data** in order to divide customers into groups. 

Our goal in this Notebook is to cluster our customers to get insights in:
- Increasing **revenue** (Knowing customers who present most of our revenue)
- Increasing customer **retention**
- Discovering **Trends and patterns** 
- Defining **customers at risk**

We will do **RFM Analysis** as a first step and then **combine RFM with predictive algorithms (k-means)**. 

RFM Analysis answers these questions:
- Who are our best customers?
- Who has the potential to be converted in more profitable customers?
- Which customers we must retain?
- Which group of customers is most likely to respond to our current campaign?

More about RFM [here](https://www.putler.com/rfm-analysis/).

In [11]:
# seting up work directory 

In [1]:
pwd

'C:\\Users\\Deepu\\Documents\\machine_learning'

In [2]:
cd C:\\Users\\Deepu\\Downloads 

C:\Users\Deepu\Downloads


In [6]:
import pandas as pd
import numpy as np

In [117]:
# load the dataset
retail_df = pd.read_excel('Online_Retail.xlsx')

In [7]:
retail_df

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country  
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
...                    ...        ...         ...             ...  
541904 2011-12-09 12:50:00       0.85     12680.0          France  
541905 2011-12-09 12:50:00       2.10     12680.0          France  
541906 2011-12-09 12:50:00       4.15     12680.0          France  
541907 2011-12-09 12:50:00       4.15     12680.0          France  
541908 2011-12-09 12:50:00       4.95     12680.0          France  

[541909 rows x 8 columns]

# Exploratory Data Analysis 

In [10]:
print("The shape of the dataset is",retail_df.shape)

The shape of the dataset is (541909, 8)


In [12]:
# check the head of the dataset 
retail_df.head(10)

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
5    536365     22752         SET 7 BABUSHKA NESTING BOXES         2   
6    536365     21730    GLASS STAR FROSTED T-LIGHT HOLDER         6   
7    536366     22633               HAND WARMER UNION JACK         6   
8    536366     22632            HAND WARMER RED POLKA DOT         6   
9    536367     84879        ASSORTED COLOUR BIRD ORNAMENT        32   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
5 2010-12-01 08:26:00       7.65     17850.0  United Kingdom  
6 2010-12-01 08:26:00       4.25     17850.0  United Kingdom  
7 2010-12-01 08:28:00       1.85     17850.0  United Kingdom  
8 2010-12-01 08:28:00       1.85     17850.0  United Kingdom  
9 2010-12-01 08:34:00       1.69     13047.0  United Kingdom

In [13]:
# check the tail of the dataset
retail_df.tail(10)

InvoiceNo StockCode                      Description  Quantity  \
541899    581587     22726       ALARM CLOCK BAKELIKE GREEN         4   
541900    581587     22730       ALARM CLOCK BAKELIKE IVORY         4   
541901    581587     22367  CHILDRENS APRON SPACEBOY DESIGN         8   
541902    581587     22629              SPACEBOY LUNCH BOX         12   
541903    581587     23256      CHILDRENS CUTLERY SPACEBOY          4   
541904    581587     22613      PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899     CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254    CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255  CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138    BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID Country  
541899 2011-12-09 12:50:00       3.75     12680.0  France  
541900 2011-12-09 12:50:00       3.75     12680.0  France  
541901 2011-12-09 12:50:00       1.95     12680.0  France  
541902 2011-12-09 12:50:00       1.95     12680.0  France  
541903 2011-12-09 12:50:00       4.15     12680.0  France  
541904 2011-12-09 12:50:00       0.85     12680.0  France  
541905 2011-12-09 12:50:00       2.10     12680.0  France  
541906 2011-12-09 12:50:00       4.15     12680.0  France  
541907 2011-12-09 12:50:00       4.15     12680.0  France  
541908 2011-12-09 12:50:00       4.95     12680.0  France

In [14]:
retail_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


The shape of the dataset is (541909,8) which means we have 541909 rows and 8 columns and by going through the above code we can say that we have null values in our dataset particularly in columns Description, CustomerID

Since we have datatype as object in non null columns and pandas just labels a column objects which has multiple data type so we might have places where we have a emply stirng let's check that as well

In [143]:
retail_df.InvoiceNo.eq('').value_counts()[0]

C:\Users\Deepu\AppData\Local\Temp\ipykernel_3048\2489092192.py:1: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  retail_df.InvoiceNo.eq('').value_counts()[0]


541909

In [150]:
# let's check if we have empty space in the object columns 

non_null_object_columns=[column for column in retail_df.describe(include = ['object']).columns.to_list() if retail_df[column]\
                         .isnull().value_counts().to_list()[0] == retail_df.shape[0]] 

#loop and check if we have empty space in the above list of non null columns which are object datatype 

boolean_list = []
for col in non_null_object_columns:
    bool_val=retail_df[col].eq('').value_counts()[0] == retail_df.shape[0]#| retail_df[col].str.isspace().value_counts())
    boolean_list.append(bool_val)

#check if we have a False contion for above loop case

if False in boolean_list:
    print('We have a empty string in one of the columns and further investigation is required')
else:
    print("we don't have empty space in the non null object columns")


we don't have empty space in the non null object columns


C:\Users\Deepu\AppData\Local\Temp\ipykernel_3048\1645573337.py:10: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  bool_val=retail_df[col].eq('').value_counts()[0] == retail_df.shape[0]#| retail_df[col].str.isspace().value_counts())
